homeassistant/components/shelly/number.py

```diff
@@ -0,0 +1,116 @@
+"""Sensor for Shelly."""
+from __future__ import annotations
+
+import asyncio
+import logging
+from typing import Any, Final, cast
+
+import async_timeout
+
+from homeassistant.components.number import NumberEntity, NumberMode
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import PERCENTAGE
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity import EntityCategory
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import AIOSHELLY_DEVICE_TIMEOUT_SEC, CONF_SLEEP_PERIOD, NUMBER_ENTITIES
+from .entity import (
+    BlockAttributeDescription,
+    ShellySleepingBlockAttributeEntity,
+    async_setup_entry_attribute_entities,
+)
+from .utils import get_device_entry_gen
+
+_LOGGER: Final = logging.getLogger(__name__)
+
+SENSORS: Final = {
+    ("device", "valvePos"): BlockAttributeDescription(
+        icon="mdi:pipe-valve",
+        name="Valve Position",
+        unit=PERCENTAGE,
+        available=lambda block: cast(int, block.valveError) != 1,
+        entity_category=EntityCategory.CONFIG,
+    ),
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up sensors for device."""
+    if get_device_entry_gen(config_entry) == 2:
+        return
+
+    if config_entry.data[CONF_SLEEP_PERIOD]:
+        await async_setup_entry_attribute_entities(
+            hass, config_entry, async_add_entities, SENSORS, BlockSleepingNumber
+        )
+
+
+class BlockSleepingNumber(ShellySleepingBlockAttributeEntity, NumberEntity):
+    """Represent a block sleeping number."""
+
+    @property
+    def value(self) -> float:
+        """Return value of number."""
+        if self.block is not None:
+            return cast(float, self.attribute_value)
+
+        return cast(float, self.last_state)
+
+    @property
+    def unit_of_measurement(self) -> str | None:
+        """Return unit of number."""
+        return cast(str, self._unit)
+
+    @property
+    def min_value(self) -> float:
+        """Return minimum value."""
+        return cast(float, NUMBER_ENTITIES[self.wrapper.model]["min"])
```

Comment:
`NUMBER_ENTITIES[self.wrapper.model]["min"]`
This doesn't look good and will not work when we want to add numbers for other devices, I don't think the entity should have hardcoded properties by device models. How will it work when we want to add another number entity for the same device?

Reply:
Fixed